pages/blog/the-anatomy-of-a-graphql-request.mdx

```diff
@@ -0,0 +1,341 @@
+This procedure is performed by the `execute` and `subscribe` functions that are exported from `graphql-js`. Other languages that are orientate themselves on the `graphql-js` reference implementation have a corresponding counterpart.
+
+# Why use GraphQL Helix?
+
+Implementing the logic for parsing and normalizing the HTTP request is time consuming and an error prone process. By leveraging a well maintained and tested library that follows the GraphQL over HTTP specification we can reduce the time required for building a GraphQL HTTP server. Furthermore, it allows us to customize and replace the functions used for doing the `parse`, `validate` and `execute`/`subscribe` phases. It is a light-weight wrapper around `graphql-js` with maximum flexibility.
+
+# Why do we want to override different phases?
+
+## Parsing with `parse`
+
+### Add Caching
+
+Parsing a GraphQL document string comes with overhead. We could cache frequently sent document strings and serve the document from the cache instead of parsing it every single time.
+
+### Test new Functionality
+
+The GraphQL Type system defines the capabilities of the GraphQL service.
+This phase can be used to add new capabilities to the type system that may not yet be supported by GraphQL specification.
+
+## Validating with `validate`
+
+### Add caching
+
+Similar to parsing, validating a GraphQL document AST comes with an overhead. We could cache recurring document ASTs and server the validation result from the cache instead of validating it every single time.
+
+### Add custom rules
+
+You might wanna restrict what kind of operations are allowed to be executed. E.g. If we we only want to allow `query` operations, we can provide a custom validation rule that yields errors as soon as a `mutation` or `subscription` operation is encountered.
+
+## Executing/Subscribing with `execute` and `subscribe`
+
+### Add caching
+
+We can serve frequently executed GraphQL operation results from a cache instead of calling all our resolvers and fetching from a remote database/server every time.
+
+### Add tracing information
+
+We can collect statistics by measuring how long it takes to resolve each field in our documents selection set and narrow down bottle-necks.
+
+### Mask and report errors
+
+We wanna make sure the errors occurring during the execution do not contain sensitive information and are properly reported, so they do not go unnoticed and are properly reported.
+
+### Add new functionality
+
+We could customize the algorithm that is used by `graphql-js` in order to add new features or make it more performant e.g. by using [graphql-executor](https://www.npmjs.com/package/graphql-executor).
+
+# Why use Envelop?
+
+While making our GraphQL servers production ready and working with our clients we discovered that we were writing the same custom code over and over again.
+
+Envelop provides us a user friendly way of hooking into the before and after phases of `parse`, `validate`, `execute` and `subscribe`.
```

Comment:
Same, a code example would be nice here to illustrate what is Envelop.

Reply:
I added one :)